File: examples/openai-agents-sds-mcp-list/src/inventory.mjs
import { REQUIRED_TOOLS } from "./constants.mjs";
import { InventoryError } from "./errors.mjs";

function fail(message, field = null) {
  throw new InventoryError(message, { field });
}

function schemaObject(tool, key) {
  const value = tool?.[key];
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    fail(`${tool?.name || "tool"} missing ${key}`, key);
  }
  return value;
}

export function assertUnpaidListInventory(tools) {
  if (!Array.isArray(tools)) fail("tools/list must return an array", "tools");
  const names = tools.map((tool) => tool?.name);
  const found = {};

  for (const required of REQUIRED_TOOLS) {
    const matches = tools.filter((tool) => tool?.name === required);
    if (matches.length !== 1) {
      fail(`tools/list missing or duplicate ${required}`, required);
    }
    const tool = matches[0];
    const inputSchema = schemaObject(tool, "inputSchema");
    if (inputSchema.type !== "object") {
      fail(`${required} inputSchema.type must be object`, required);
    }
    found[required] = tool;
  }

  const extractInput = found.extract.inputSchema;
  if (!extractInput.properties || typeof extractInput.properties !== "object") {
    fail("extract inputSchema.properties is required", "extract");
  }
  if (!extractInput.properties.url) {
    fail("extract inputSchema must include url", "extract");
  }
  if (!Array.isArray(extractInput.required) || !extractInput.required.includes("url")) {
    fail("extract inputSchema.required must include url", "extract");
  }

  const batchInput = found.extract_batch.inputSchema;
  if (!batchInput.properties || typeof batchInput.properties !== "object") {
    fail("extract_batch inputSchema.properties is required", "extract_batch");
  }
  if (!batchInput.properties.urls) {
    fail("extract_batch inputSchema must include urls", "extract_batch");
  }
  if (!Array.isArray(batchInput.required) || !batchInput.required.includes("urls")) {
    fail("extract_batch inputSchema.required must include urls", "extract_batch");
  }

  return {
    names,
    extract: found.extract,
    extract_batch: found.extract_batch,
    extractPresent: true,
    extractBatchPresent: true,
    extractInputRequiresUrl: true,
    extractBatchInputRequiresUrls: true,
    exactGlobalCountRequired: false,
  };
}

File: examples/crewai-sds-mcp-list/src/crewai_sds_mcp_list/inventory.py
"""Require extract and extract_batch on an unpaid tools/list. Extra tools are fine."""

from __future__ import annotations

from .constants import MAX_TOOL_COUNT, REQUIRED_TOOLS
from .errors import InventoryError
from .policy import tool_name


def _input_properties(tool: dict) -> dict:
    schema = tool.get("inputSchema")
    if not isinstance(schema, dict):
        return {}
    properties = schema.get("properties")
    return properties if isinstance(properties, dict) else {}


def assert_unpaid_sds_inventory(tools) -> dict:
    if not isinstance(tools, list):
        raise InventoryError("tools/list must return a list")
    if len(tools) > MAX_TOOL_COUNT:
        raise InventoryError(f"tools/list exceeds the tool ceiling of {MAX_TOOL_COUNT}")

    by_name: dict[str, dict] = {}
    names: list[str] = []
    for tool in tools:
        if not isinstance(tool, dict):
            raise InventoryError("tools/list contains a non-object tool")
        name = tool_name(tool)
        if not name:
            raise InventoryError("tools/list contains a tool with no name")
        if name in by_name:
            raise InventoryError(f"tools/list duplicate name: {name}")
        by_name[name] = tool
        names.append(name)

    missing = [name for name in REQUIRED_TOOLS if name not in by_name]
    if missing:
        raise InventoryError(
            "tools/list missing required tools: " + ", ".join(missing)
        )

    extract_props = _input_properties(by_name["extract"])
    if "url" not in extract_props:
        raise InventoryError("extract inputSchema is missing url")

    batch_props = _input_properties(by_name["extract_batch"])
    if "urls" not in batch_props:
        raise InventoryError("extract_batch inputSchema is missing urls")

    return {
        "extract": by_name["extract"],
        "extract_batch": by_name["extract_batch"],
        "names": names,
        "by_name": by_name,
    }
